Add validation layer for Derefer

ouz-a opened the pull request:
_Follow up work to #96549 #96116 #95857 #95649_

This adds validation for Derefer making sure it is always the first projection.

r? rust-lang/mir-opt

ouz-a committed:
validate derefer, run derefer inside generator

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6428146073?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/e2492b08968faaf4f827256d38d9448d5a69713a/checks/6428146073/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling rustc_query_impl v0.0.0 (/checkout/compiler/rustc_query_impl)
   Compiling rustc_monomorphize v0.0.0 (/checkout/compiler/rustc_monomorphize)
   Compiling rustc_passes v0.0.0 (/checkout/compiler/rustc_passes)
   Compiling rustc_save_analysis v0.0.0 (/checkout/compiler/rustc_save_analysis)
thread 'rustc' panicked at '(*(_1.0: &mut [generator@compiler/rustc_metadata/src/rmeta/encoder.rs:1136:73: 1155:14]))', compiler/rustc_const_eval/src/transform/validate.rs:314:13
stack backtrace:
   0:     0x7f596fbd0da2 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::ha288dc719902e24d
   1:     0x7f596fc386c8 - core::fmt::write::h42234c3e51154f4c
   2:     0x7f596fbc10f1 - std::io::Write::write_fmt::h74fbb9643da2d185
   3:     0x7f596fbd4106 - std::panicking::default_hook::{{closure}}::h3bfe018301273550
   4:     0x7f596fbd3cfd - std::panicking::default_hook::h4173afa32faa81d9
   5:     0x7f59707288e1 - rustc_driver[1a777b00ee158e36]::DEFAULT_HOOK::{closure#0}::{closure#0}
   6:     0x7f596fbd4aa0 - std::panicking::rust_panic_with_hook::h59cc3082e9104592
   7:     0x7f596fbd48b7 - std::panicking::begin_panic_handler::{{closure}}::h79b0ac1d2b9c8b15
   8:     0x7f596fbd1344 - std::sys_common::backtrace::__rust_end_short_backtrace::h72d3f8515fc7966b
   9:     0x7f596fbd45a9 - rust_begin_unwind
  10:     0x7f596fb880b3 - core::panicking::panic_fmt::h0eedb34d228802aa
  11:     0x7f5971c942cb - <rustc_const_eval[463eeed51a022010]::transform::validate::TypeChecker as rustc_middle[8d4dc3708b593ac1]::mir::visit::Visitor>::visit_place
  12:     0x7f5971c9528d - <rustc_const_eval[463eeed51a022010]::transform::validate::TypeChecker as rustc_middle[8d4dc3708b593ac1]::mir::visit::Visitor>::visit_rvalue
  13:     0x7f5971c9581a - <rustc_const_eval[463eeed51a022010]::transform::validate::TypeChecker as rustc_middle[8d4dc3708b593ac1]::mir::visit::Visitor>::visit_statement
  14:     0x7f5971c92c2f - <rustc_const_eval[463eeed51a022010]::transform::validate::Validator as rustc_middle[8d4dc3708b593ac1]::mir::MirPass>::run_pass
  15:     0x7f5970cc1867 - rustc_mir_transform[8cd1ea75711a0041]::pass_manager::run_passes
  16:     0x7f5970dace7b - rustc_mir_transform[8cd1ea75711a0041]::optimized_mir
  17:     0x7f5971e2eead - rustc_query_system[4cf300c5f65a2dfc]::query::plumbing::try_execute_query::<rustc_query_impl[b8302d1c2e1c865b]::plumbing::QueryCtxt, rustc_query_system[4cf300c5f65a2dfc]::query::caches::DefaultCache<rustc_span[5c736203a6ab5594]::def_id::DefId, &rustc_middle[8d4dc3708b593ac1]::mir::Body>>
  18:     0x7f5971ecbe42 - rustc_query_system[4cf300c5f65a2dfc]::query::plumbing::get_query::<rustc_query_impl[b8302d1c2e1c865b]::queries::optimized_mir, rustc_query_impl[b8302d1c2e1c865b]::plumbing::QueryCtxt>
  19:     0x7f5972270ad9 - <rustc_query_impl[b8302d1c2e1c865b]::Queries as rustc_middle[8d4dc3708b593ac1]::ty::query::QueryEngine>::optimized_mir
  20:     0x7f59726139fd - <rustc_metadata[6387b7782e9ee88c]::rmeta::encoder::EncodeContext>::encode_crate_root
  21:     0x7f597262874b - rustc_metadata[6387b7782e9ee88c]::rmeta::encoder::encode_metadata_impl
  22:     0x7f59726ed831 - rustc_data_structures[db21f27220b58c22]::sync::join::<rustc_metadata[6387b7782e9ee88c]::rmeta::encoder::encode_metadata::{closure#0}, rustc_metadata[6387b7782e9ee88c]::rmeta::encoder::encode_metadata::{closure#1}, rustc_metadata[6387b7782e9ee88c]::rmeta::encoder::EncodedMetadata, ()>
  23:     0x7f5972627d4e - rustc_metadata[6387b7782e9ee88c]::rmeta::encoder::encode_metadata
  24:     0x7f5970838c7c - <rustc_interface[b11680d0c9f2606a]::passes::QueryContext>::enter::<<rustc_interface[b11680d0c9f2606a]::queries::Queries>::ongoing_codegen::{closure#0}::{closure#0}, core[10878fb91fc84a80]::result::Result<alloc[24f448636cd10183]::boxed::Box<dyn core[10878fb91fc84a80]::any::Any>, rustc_errors[984494b0cf0e650]::ErrorGuaranteed>>
  25:     0x7f597081bbfe - <rustc_interface[b11680d0c9f2606a]::queries::Queries>::ongoing_codegen
  26:     0x7f59706be378 - <rustc_interface[b11680d0c9f2606a]::interface::Compiler>::enter::<rustc_driver[1a777b00ee158e36]::run_compiler::{closure#1}::{closure#2}, core[10878fb91fc84a80]::result::Result<core[10878fb91fc84a80]::option::Option<rustc_interface[b11680d0c9f2606a]::queries::Linker>, rustc_errors[984494b0cf0e650]::ErrorGuaranteed>>
  27:     0x7f59706a0306 - rustc_span[5c736203a6ab5594]::with_source_map::<core[10878fb91fc84a80]::result::Result<(), rustc_errors[984494b0cf0e650]::ErrorGuaranteed>, rustc_interface[b11680d0c9f2606a]::interface::create_compiler_and_run<core[10878fb91fc84a80]::result::Result<(), rustc_errors[984494b0cf0e650]::ErrorGuaranteed>, rustc_driver[1a777b00ee158e36]::run_compiler::{closure#1}>::{closure#1}>
  28:     0x7f59706bf60f - <scoped_tls[5ed4b67c3b198af5]::ScopedKey<rustc_span[5c736203a6ab5594]::SessionGlobals>>::set::<rustc_interface[b11680d0c9f2606a]::interface::run_compiler<core[10878fb91fc84a80]::result::Result<(), rustc_errors[984494b0cf0e650]::ErrorGuaranteed>, rustc_driver[1a777b00ee158e36]::run_compiler::{closure#1}>::{closure#0}, core[10878fb91fc84a80]::result::Result<(), rustc_errors[984494b0cf0e650]::ErrorGuaranteed>>
  29:     0x7f59707142a9 - std[eba90a372f7a1edd]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[b11680d0c9f2606a]::util::run_in_thread_pool_with_globals<rustc_interface[b11680d0c9f2606a]::interface::run_compiler<core[10878fb91fc84a80]::result::Result<(), rustc_errors[984494b0cf0e650]::ErrorGuaranteed>, rustc_driver[1a777b00ee158e36]::run_compiler::{closure#1}>::{closure#0}, core[10878fb91fc84a80]::result::Result<(), rustc_errors[984494b0cf0e650]::ErrorGuaranteed>>::{closure#0}, core[10878fb91fc84a80]::result::Result<(), rustc_errors[984494b0cf0e650]::ErrorGuaranteed>>
  30:     0x7f59706d3051 - std[eba90a372f7a1edd]::panicking::try::<core[10878fb91fc84a80]::result::Result<(), rustc_errors[984494b0cf0e650]::ErrorGuaranteed>, core[10878fb91fc84a80]::panic::unwind_safe::AssertUnwindSafe<<std[eba90a372f7a1edd]::thread::Builder>::spawn_unchecked_<rustc_interface[b11680d0c9f2606a]::util::run_in_thread_pool_with_globals<rustc_interface[b11680d0c9f2606a]::interface::run_compiler<core[10878fb91fc84a80]::result::Result<(), rustc_errors[984494b0cf0e650]::ErrorGuaranteed>, rustc_driver[1a777b00ee158e36]::run_compiler::{closure#1}>::{closure#0}, core[10878fb91fc84a80]::result::Result<(), rustc_errors[984494b0cf0e650]::ErrorGuaranteed>>::{closure#0}, core[10878fb91fc84a80]::result::Result<(), rustc_errors[984494b0cf0e650]::ErrorGuaranteed>>::{closure#1}::{closure#0}>>
  31:     0x7f5970716522 - <<std[eba90a372f7a1edd]::thread::Builder>::spawn_unchecked_<rustc_interface[b11680d0c9f2606a]::util::run_in_thread_pool_with_globals<rustc_interface[b11680d0c9f2606a]::interface::run_compiler<core[10878fb91fc84a80]::result::Result<(), rustc_errors[984494b0cf0e650]::ErrorGuaranteed>, rustc_driver[1a777b00ee158e36]::run_compiler::{closure#1}>::{closure#0}, core[10878fb91fc84a80]::result::Result<(), rustc_errors[984494b0cf0e650]::ErrorGuaranteed>>::{closure#0}, core[10878fb91fc84a80]::result::Result<(), rustc_errors[984494b0cf0e650]::ErrorGuaranteed>>::{closure#1} as core[10878fb91fc84a80]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  32:     0x7f596fbe1403 - std::sys::unix::thread::Thread::new::thread_start::h6bc2e8f9e4e3d29f
  33:     0x7f596a131609 - start_thread
  34:     0x7f596fa44163 - clone
  35:                0x0 - <unknown>
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.


note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.62.0-nightly (8290b4cb6 2022-05-13) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type lib -C opt-level=3 -C embed-bitcode=no -C debuginfo=0 -C debug-assertions=on -Z unstable-options -C symbol-mangling-version=v0 -Z unstable-options -Z macro-backtrace -C link-args=-Wl,-z,origin -C link-args=-Wl,-rpath,$ORIGIN/../lib -Z unstable-options -C split-debuginfo=off -Z tls-model=initial-exec -Z unstable-options -C prefer-dynamic -C llvm-args=-import-instr-limit=10 -Z binary-dep-depinfo -Z force-unstable-if-unmarked
note: some of the compiler flags provided by cargo are hidden

query stack during panic:
query stack during panic:
#0 [optimized_mir] optimizing MIR for `rmeta::encoder::<impl at compiler/rustc_metadata/src/rmeta/encoder.rs:987:1: 1964:2>::encode_info_for_mod::{closure#0}`
error: could not compile `rustc_metadata`
warning: build failed, waiting for other jobs to finish...
Build completed unsuccessfully in 0:08:15
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6429372523?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/ddd58edcfd7fcf2b303b52f61a9fe705688c9d68/checks/6429372523/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Check compiletest suite=mir-opt mode=mir-opt (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 174 tests
......................................i................................................. 88/174
Some tests failed in compiletest suite=mir-opt mode=mir-opt host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
..i...F..............................................i................................

---- [mir-opt] src/test/mir-opt/inline/inline-generator.rs stdout ----
---- [mir-opt] src/test/mir-opt/inline/inline-generator.rs stdout ----
30 +         let mut _10: bool;               // in scope 6 at $DIR/inline-generator.rs:15:9: 15:9
31 +         let _11: bool;                   // in scope 6 at $DIR/inline-generator.rs:15:6: 15:7
32 +         let mut _12: u32;                // in scope 6 at $DIR/inline-generator.rs:15:5: 15:41
+ +         let mut _13: &mut [generator@$DIR/inline-generator.rs:15:5: 15:41]; // in scope 6 at $DIR/inline-generator.rs:15:5: 15:41
+ +         let mut _14: &mut [generator@$DIR/inline-generator.rs:15:5: 15:41]; // in scope 6 at $DIR/inline-generator.rs:15:5: 15:41
+ +         let mut _15: &mut [generator@$DIR/inline-generator.rs:15:5: 15:41]; // in scope 6 at $DIR/inline-generator.rs:15:5: 15:41
34   
35       bb0: {


73 +         StorageLive(_10);                // scope 0 at $DIR/inline-generator.rs:9:14: 9:46
74 +         StorageLive(_11);                // scope 0 at $DIR/inline-generator.rs:9:14: 9:46
75 +         StorageLive(_12);                // scope 0 at $DIR/inline-generator.rs:9:14: 9:46
- +         _12 = discriminant((*(_2.0: &mut [generator@$DIR/inline-generator.rs:15:5: 15:41]))); // scope 6 at $DIR/inline-generator.rs:15:5: 15:41
+ +         StorageLive(_13);                // scope 6 at $DIR/inline-generator.rs:15:5: 15:41
+ +         _13 = move (_2.0: &mut [generator@$DIR/inline-generator.rs:15:5: 15:41]); // scope 6 at $DIR/inline-generator.rs:15:5: 15:41
+ +         _12 = discriminant((*_13));      // scope 6 at $DIR/inline-generator.rs:15:5: 15:41
+ +         StorageDead(_13);                // scope 6 at $DIR/inline-generator.rs:15:5: 15:41
77 +         switchInt(move _12) -> [0_u32: bb3, 1_u32: bb8, 3_u32: bb7, otherwise: bb9]; // scope 6 at $DIR/inline-generator.rs:15:5: 15:41
79   


118 +         Deinit(_1);                      // scope 6 at $DIR/inline-generator.rs:15:11: 15:39
119 +         ((_1 as Yielded).0: i32) = move _8; // scope 6 at $DIR/inline-generator.rs:15:11: 15:39
120 +         discriminant(_1) = 0;            // scope 6 at $DIR/inline-generator.rs:15:11: 15:39
- +         discriminant((*(_2.0: &mut [generator@$DIR/inline-generator.rs:15:5: 15:41]))) = 3; // scope 6 at $DIR/inline-generator.rs:15:11: 15:39
+ +         StorageLive(_14);                // scope 6 at $DIR/inline-generator.rs:15:11: 15:39
+ +         _14 = move (_2.0: &mut [generator@$DIR/inline-generator.rs:15:5: 15:41]); // scope 6 at $DIR/inline-generator.rs:15:11: 15:39
+ +         discriminant((*_14)) = 3;        // scope 6 at $DIR/inline-generator.rs:15:11: 15:39
+ +         StorageDead(_14);                // scope 6 at $DIR/inline-generator.rs:15:11: 15:39
122 +         goto -> bb1;                     // scope 0 at $DIR/inline-generator.rs:15:11: 15:39
124 + 


129 +         Deinit(_1);                      // scope 6 at $DIR/inline-generator.rs:15:41: 15:41
130 +         ((_1 as Complete).0: bool) = move _10; // scope 6 at $DIR/inline-generator.rs:15:41: 15:41
131 +         discriminant(_1) = 1;            // scope 6 at $DIR/inline-generator.rs:15:41: 15:41
- +         discriminant((*(_2.0: &mut [generator@$DIR/inline-generator.rs:15:5: 15:41]))) = 1; // scope 6 at $DIR/inline-generator.rs:15:41: 15:41
+ +         StorageLive(_15);                // scope 6 at $DIR/inline-generator.rs:15:41: 15:41
+ +         _15 = move (_2.0: &mut [generator@$DIR/inline-generator.rs:15:5: 15:41]); // scope 6 at $DIR/inline-generator.rs:15:41: 15:41
+ +         discriminant((*_15)) = 1;        // scope 6 at $DIR/inline-generator.rs:15:41: 15:41
+ +         StorageDead(_15);                // scope 6 at $DIR/inline-generator.rs:15:41: 15:41
133 +         goto -> bb1;                     // scope 0 at $DIR/inline-generator.rs:15:41: 15:41
135 + 


thread '[mir-opt] src/test/mir-opt/inline/inline-generator.rs' panicked at 'Actual MIR output differs from expected MIR output /checkout/src/test/mir-opt/inline/inline_generator.main.Inline.diff', src/tools/compiletest/src/runtest.rs:3410:25


failures:
    [mir-opt] src/test/mir-opt/inline/inline-generator.rs
```

</details>

oli-obk commented on compiler/rustc_mir_transform/src/deref_separator.rs:
```diff
@@ -1,11 +1,104 @@
+    }
+}
+
+pub fn mini_deref_finder<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
```
What's the difference between this one and the main one? Can we deduplicate their logic if the differences are small?

ouz-a replied:
I added two comments explaining it. I think running `deref_finder` two times is a bit expensive, the new function is a bit less expensive compared to the old one.

oli-obk replied:
Please just run the full version. We'll benchmark it and if there's a regression we can look into it

ouz-a commented on compiler/rustc_mir_transform/src/deref_separator.rs:
```diff
@@ -1,11 +1,104 @@
+    }
+
+    fn visit_place(&mut self, place: &mut Place<'tcx>, cntxt: PlaceContext, loc: Location) {
+        if place.projection.len() > 1
```
These checks can save a lot of unnecessary iterations.

oli-obk replied:
I think the full version should just have that, too

ouz-a replied:
That makes a lot more sense 😅

ouz-a commented on compiler/rustc_mir_transform/src/deref_separator.rs:
```diff
@@ -1,11 +1,104 @@
+            }
+
+            for (idx, (p_ref, _p_elem)) in place.iter_projections().enumerate() {
+                if last_deref_idx > 0 {
```
Although it's small, check here is lighter compared to other function.

ouz-a commented on compiler/rustc_mir_transform/src/deref_separator.rs:
```diff
@@ -8,23 +8,21 @@ use rustc_middle::ty::TyCtxt;
     fn visit_place(&mut self, place: &mut Place<'tcx>, cntxt: PlaceContext, loc: Location) {
-        if place.projection.len() > 1
-            && place.projection[1..].contains(&ProjectionElem::Deref)
+        if !place.projection.is_empty()
```
I think this should remove a lot of unnecessary computation

oli-obk commented on compiler/rustc_mir_transform/src/deref_separator.rs:
```diff
@@ -17,63 +19,68 @@ impl<'tcx> MutVisitor<'tcx> for DerefChecker<'tcx> {
+    fn visit_place(&mut self, place: &mut Place<'tcx>, cntxt: PlaceContext, loc: Location) {
+        if !place.projection.is_empty()
+            && cntxt != PlaceContext::NonUse(VarDebugInfo)
+            && place.projection[0..].contains(&ProjectionElem::Deref)
```
```suggestion
            && place.projection[1..].contains(&ProjectionElem::Deref)
```

oli-obk commented on compiler/rustc_mir_transform/src/deref_separator.rs:
```diff
@@ -17,63 +19,68 @@ impl<'tcx> MutVisitor<'tcx> for DerefChecker<'tcx> {
-                    let temp_place =
-                        Place::from(temp).project_deeper(&place.projection[idx..], self.tcx);
-                    *place = temp_place;
+            for (idx, (_p_ref, p_elem)) in place.iter_projections().enumerate() {
```
This can just iterate over the projection slice instead of iter_projections

ouz-a commented on compiler/rustc_mir_transform/src/deref_separator.rs:
```diff
@@ -22,16 +22,16 @@ impl<'tcx> MutVisitor<'tcx> for DerefChecker<'tcx> {
 
-            for (idx, (_p_ref, p_elem)) in place.iter_projections().enumerate() {
-                if p_elem == ProjectionElem::Deref {
+            for (idx,_) in place.projection[0..].iter().enumerate() {
```
I am not sure if this change is going to affect anything at all

oli-obk commented on compiler/rustc_mir_transform/src/deref_separator.rs:
```diff
@@ -17,63 +19,68 @@ impl<'tcx> MutVisitor<'tcx> for DerefChecker<'tcx> {
-                        Place::from(temp).project_deeper(&place.projection[idx..], self.tcx);
-                    *place = temp_place;
+            for (idx, _) in place.projection[0..].iter().enumerate() {
+                if place.projection[idx] == ProjectionElem::Deref {
```
Instead of indexing here use the second for loop value

oli-obk commented:
re-rolling as I wrote the change adding a new mir phase (cc @JakobDegen)

r? rust-lang/mir-opt

oli-obk commented:
r? @JakobDegen

oli-obk commented:
r? @davidtwco

oli-obk commented:
@bors try

We should crater this as it adds a new rule to our final MIR output that needs to be upheld or we ICE

bors commented:
:hourglass: Trying commit 15fb808df52dedb6cd791a19a8eaa3ad88855ba8 with merge 3930b75493eb88cd5431103dbdbc47eb760e9d9b...
<!-- homu: {"type":"TryBuildStarted","head_sha":"15fb808df52dedb6cd791a19a8eaa3ad88855ba8","merge_sha":"3930b75493eb88cd5431103dbdbc47eb760e9d9b"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/6458178583?check_suite_focus=true)
Build commit: 3930b75493eb88cd5431103dbdbc47eb760e9d9b (`3930b75493eb88cd5431103dbdbc47eb760e9d9b`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/6458178583?check_suite_focus=true"},"merge_sha":"3930b75493eb88cd5431103dbdbc47eb760e9d9b"} -->

JakobDegen commented on compiler/rustc_const_eval/src/transform/validate.rs:
```diff
@@ -301,9 +302,16 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
+            && place.projection.len() > 1
+            && cntxt != PlaceContext::NonUse(VarDebugInfo)
+        {
+            assert!(!place.projection[1..].contains(&ProjectionElem::Deref), "{place:#?}");
```
Is there any reason not to use `self.fail` here?

ouz-a replied:
It gave horrendous wall of text as error I thought assert's error looked cleaner and concise.

JakobDegen replied:
Hmm, honestly, I'm not so concerned by the wall of text. This is an ICE after all, it doesn't really have to have good UX, just needs to have all the necessary information. Especially the part about the location being included in the ICE seems really useful.

ouz-a replied:
I changed it to `self.fail` after trying it out it is indeed more useful than basic assert

oli-obk commented:
@craterbot run mode=build-only

need to run in build mode so that we actually run all mir optimizations

Mark-Simulacrum commented:
@oli-obk just to confirm - debug build is enough to run MIR opts? I don't think crater builds with the release profile today by default.

oli-obk commented:
> debug build is enough to run MIR opts? I don't think crater builds with the release profile today by default.

yes, it's not important to run any mir opts, it's just important to produce MIR for LLVM

davidtwco reviewed (approved):
LGTM, r=me after resolving other comments

JakobDegen commented:
Dest prop is affected by this actually, the version in my PR handles it correctly though

oli-obk commented:
crater is clean. So I guess what's left is adding the inliner test and fixing it by running derefer at the end of the inliner

ouz-a commented:
> There are a few transformation that replace all occurrences of a local with a place. Do they need to be updated if the replacement place might contain any projections, or is it an issue only if projections contain a dereference?
> 
>     * State transform - it seems it is already updated. Ok.
> 
>     * NRVO - replaces only complete locals. Ok.
> 
>     * Inliner - replaces locals with arbitrary places. Fails the new validation. Disabled by default, but a pull requests enabling it waits on bors.
> 
>     * Destination propagation - a replacement can contain projections, but not deref projection. Disabled by default.
> 
> 
> MIR inliner test case
> 
> ```
> $ cat a.rs 
> #![feature(box_syntax)]
> #[inline]
> fn f() -> Box<i32> { box 0 }
> fn main() { box f(); }
> $ rustc +3930b75493eb88cd5431103dbdbc47eb760e9d9b a.rs -Zinline-mir
> thread 'rustc' panicked at '(*(*_7))', compiler/rustc_const_eval/src/transform/validate.rs:314:13
> ```

Thanks, I added this as test and fixed it by running derefer at the end of inline

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6589864782?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/6714c645bf9758f763e0c9c5cfa58b2a0cc4d3bd/checks/6589864782/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 finished in 0.529 seconds
Check compiletest suite=mir-opt mode=mir-opt (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 176 tests
i......................................i............................................F.F. 88/176
....i......F...........................................i................................ 176/176
Some tests failed in compiletest suite=mir-opt mode=mir-opt host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
failures:

---- [mir-opt] src/test/mir-opt/inline/dyn-trait.rs stdout ----
---- [mir-opt] src/test/mir-opt/inline/dyn-trait.rs stdout ----
32 +         StorageDead(_4);                 // scope 1 at $DIR/dyn-trait.rs:21:21: 21:22
33           StorageDead(_2);                 // scope 0 at $DIR/dyn-trait.rs:27:15: 27:16
34           return;                          // scope 0 at $DIR/dyn-trait.rs:28:2: 28:2
+ +     }
+ + 
+ +     bb2 (cleanup): {
+ +         resume;                          // scope 0 at $DIR/dyn-trait.rs:26:1: 28:2
36   }
37   


thread '[mir-opt] src/test/mir-opt/inline/dyn-trait.rs' panicked at 'Actual MIR output differs from expected MIR output /checkout/src/test/mir-opt/inline/dyn_trait.try_execute_query.Inline.diff', src/tools/compiletest/src/runtest.rs:3410:25

---- [mir-opt] src/test/mir-opt/inline/inline-cycle-generic.rs stdout ----
---- [mir-opt] src/test/mir-opt/inline/inline-cycle-generic.rs stdout ----
17           StorageDead(_1);                 // scope 0 at $DIR/inline-cycle-generic.rs:9:24: 9:25
18           _0 = const ();                   // scope 0 at $DIR/inline-cycle-generic.rs:8:11: 10:2
19           return;                          // scope 0 at $DIR/inline-cycle-generic.rs:10:2: 10:2
- +     }
- + 
- +     bb2 (cleanup): {
- +         resume;                          // scope 0 at $DIR/inline-cycle-generic.rs:8:1: 10:2
25   }
26   


thread '[mir-opt] src/test/mir-opt/inline/inline-cycle-generic.rs' panicked at 'Actual MIR output differs from expected MIR output /checkout/src/test/mir-opt/inline/inline_cycle_generic.main.Inline.diff', src/tools/compiletest/src/runtest.rs:3410:25
---- [mir-opt] src/test/mir-opt/inline/inline-compatibility.rs stdout ----
---- [mir-opt] src/test/mir-opt/inline/inline-compatibility.rs stdout ----
19           StorageDead(_1);                 // scope 0 at $DIR/inline-compatibility.rs:13:21: 13:22
20           _0 = const ();                   // scope 0 at $DIR/inline-compatibility.rs:12:40: 14:2
21           return;                          // scope 0 at $DIR/inline-compatibility.rs:14:2: 14:2
+ +     }
+ + 
+ +     bb1 (cleanup): {
+ +         resume;                          // scope 0 at $DIR/inline-compatibility.rs:12:1: 14:2
23   }
24   


thread '[mir-opt] src/test/mir-opt/inline/inline-compatibility.rs' panicked at 'Actual MIR output differs from expected MIR output /checkout/src/test/mir-opt/inline/inline_compatibility.inlined_target_feature.Inline.diff', src/tools/compiletest/src/runtest.rs:3410:25

failures:
    [mir-opt] src/test/mir-opt/inline/dyn-trait.rs
    [mir-opt] src/test/mir-opt/inline/inline-compatibility.rs
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6590247076?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/f99c0cf76c0a8781d0957d2a2404f5717991155e/checks/6590247076/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Check compiletest suite=mir-opt mode=mir-opt (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 176 tests
i......................................i................................................ 88/176
....i...F..............................................i................................ 176/176
Some tests failed in compiletest suite=mir-opt mode=mir-opt host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
failures:

---- [mir-opt] src/test/mir-opt/inline/inline-compatibility.rs stdout ----
---- [mir-opt] src/test/mir-opt/inline/inline-compatibility.rs stdout ----
19           StorageDead(_1);                 // scope 0 at $DIR/inline-compatibility.rs:13:21: 13:22
20           _0 = const ();                   // scope 0 at $DIR/inline-compatibility.rs:12:40: 14:2
21           return;                          // scope 0 at $DIR/inline-compatibility.rs:14:2: 14:2
+ +     }
+ + 
+ +     bb1 (cleanup): {
+ +         resume;                          // scope 0 at $DIR/inline-compatibility.rs:12:1: 14:2
23   }
24   


thread '[mir-opt] src/test/mir-opt/inline/inline-compatibility.rs' panicked at 'Actual MIR output differs from expected MIR output /checkout/src/test/mir-opt/inline/inline_compatibility.inlined_target_feature.Inline.diff', src/tools/compiletest/src/runtest.rs:3410:25


failures:
    [mir-opt] src/test/mir-opt/inline/inline-compatibility.rs
```

</details>

davidtwco reviewed (approved):
I'm not sure if this is waiting for my review, but if it is, r=me once CI is passing.

oli-obk commented:
@bors r=davidtwco

bors commented:
:pushpin: Commit 1c048eb5c01040feae731560f3c95f8647beb218 has been approved by `davidtwco`

<!-- @bors r=davidtwco 1c048eb5c01040feae731560f3c95f8647beb218 -->
<!-- homu: {"type":"Approved","sha":"1c048eb5c01040feae731560f3c95f8647beb218","approver":"davidtwco"} -->

GuillaumeGomez commented:
Seems like it failed in https://github.com/rust-lang/rust/pull/97496.

@bors r-

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6643801275?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/c58fc70b49bf95b6057d14aff51f9e41be322be3/checks/6643801275/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Check compiletest suite=mir-opt mode=mir-opt (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 178 tests
i......................................i................................................ 88/178
......i...F..............................................i.............................. 176/178
Some tests failed in compiletest suite=mir-opt mode=mir-opt host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
failures:

---- [mir-opt] src/test/mir-opt/inline/inline-compatibility.rs stdout ----
---- [mir-opt] src/test/mir-opt/inline/inline-compatibility.rs stdout ----
19           StorageDead(_1);                 // scope 0 at $DIR/inline-compatibility.rs:13:21: 13:22
20           _0 = const ();                   // scope 0 at $DIR/inline-compatibility.rs:12:40: 14:2
21           return;                          // scope 0 at $DIR/inline-compatibility.rs:14:2: 14:2
+ +     }
+ + 
+ +     bb1 (cleanup): {
+ +         resume;                          // scope 0 at $DIR/inline-compatibility.rs:12:1: 14:2
23   }
24   


thread '[mir-opt] src/test/mir-opt/inline/inline-compatibility.rs' panicked at 'Actual MIR output differs from expected MIR output /checkout/src/test/mir-opt/inline/inline_compatibility.inlined_target_feature.Inline.diff', src/tools/compiletest/src/runtest.rs:3420:25


failures:
    [mir-opt] src/test/mir-opt/inline/inline-compatibility.rs
```

</details>

oli-obk commented:
Please rebase and squash once everything works

oli-obk commented:
@bors r=davidtwco

bors commented:
:pushpin: Commit 6f1a1a9923f0798e695fbbd1054f1f3adbbc8820 has been approved by `davidtwco`

<!-- @bors r=davidtwco 6f1a1a9923f0798e695fbbd1054f1f3adbbc8820 -->
<!-- homu: {"type":"Approved","sha":"6f1a1a9923f0798e695fbbd1054f1f3adbbc8820","approver":"davidtwco"} -->

bors commented:
:hourglass: Testing commit 6f1a1a9923f0798e695fbbd1054f1f3adbbc8820 with merge d529f56690e100bf62285cba36ca4019ffd14ff9...
<!-- homu: {"type":"BuildStarted","head_sha":"6f1a1a9923f0798e695fbbd1054f1f3adbbc8820","merge_sha":"d529f56690e100bf62285cba36ca4019ffd14ff9"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/6646335435?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/6646335435?check_suite_focus=true","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`test-various`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/6646246808?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/d529f56690e100bf62285cba36ca4019ffd14ff9/checks/6646246808/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test [mir-opt] src/test/mir-opt/storage_live_dead_in_statics.rs ... ok

failures:

---- [mir-opt] src/test/mir-opt/derefer_inline_test.rs stdout ----
25       bb1: {
26           StorageLive(_5);                 // scope 0 at $DIR/derefer_inline_test.rs:10:5: 10:12
27           _5 = ShallowInitBox(move _4, std::boxed::Box<u32>); // scope 0 at $DIR/derefer_inline_test.rs:10:5: 10:12
-           (*_5) = f() -> [return: bb2, unwind: bb5]; // scope 0 at $DIR/derefer_inline_test.rs:10:9: 10:12
+           (*_5) = f() -> bb2;              // scope 0 at $DIR/derefer_inline_test.rs:10:9: 10:12
29                                            // mir::Constant
30                                            // + span: $DIR/derefer_inline_test.rs:10:9: 10:10
31                                            // + literal: Const { ty: fn() -> Box<u32> {f}, val: Value(Scalar(<ZST>)) }
38   
39       bb3: {
39       bb3: {
40           StorageDead(_5);                 // scope 0 at $DIR/derefer_inline_test.rs:10:11: 10:12
-           drop(_1) -> [return: bb4, unwind: bb6]; // scope 0 at $DIR/derefer_inline_test.rs:10:12: 10:13
+           drop(_1) -> bb4;                 // scope 0 at $DIR/derefer_inline_test.rs:10:12: 10:13
43   
44       bb4: {

48       }
48       }
49   
50       bb5 (cleanup): {
-           goto -> bb8;                     // scope 0 at $DIR/derefer_inline_test.rs:10:11: 10:12
-   
-   
-       bb6 (cleanup): {
55           resume;                          // scope 0 at $DIR/derefer_inline_test.rs:9:1: 11:2
-   
-   
-       bb7 (cleanup): {
-           _6 = alloc::alloc::box_free::<Box<u32>, std::alloc::Global>(move (_5.0: std::ptr::Unique<std::boxed::Box<u32>>), move (_5.1: std::alloc::Global)) -> bb6; // scope 0 at $DIR/derefer_inline_test.rs:10:11: 10:12
-                                            // mir::Constant
-                                            // + span: $DIR/derefer_inline_test.rs:10:11: 10:12
-                                            // + literal: Const { ty: unsafe fn(Unique<Box<u32>>, std::alloc::Global) {alloc::alloc::box_free::<Box<u32>, std::alloc::Global>}, val: Value(Scalar(<ZST>)) }
-   
-   
-       bb8 (cleanup): {
-           goto -> bb7;                     // scope 0 at $DIR/derefer_inline_test.rs:10:11: 10:12
68   }
69   


thread '[mir-opt] src/test/mir-opt/derefer_inline_test.rs' panicked at 'Actual MIR output differs from expected MIR output /checkout/src/test/mir-opt/derefer_inline_test.main.Derefer.diff', src/tools/compiletest/src/runtest.rs:3420:25


failures:
    [mir-opt] src/test/mir-opt/derefer_inline_test.rs
```

</details>

tmiasko commented:
The tests fail on wasm32-unknown-unknown, because this target defaults to `-C panic=abort`, and the function calls in MIR don't have unwind edges:

```diff
-           (*_5) = f() -> [return: bb2, unwind: bb5]; // scope 0 at $DIR/derefer_inline_test.rs:10:9: 10:12
+           (*_5) = f() -> bb2;
```

We have been generally either ignoring test on this target:
 `// ignore-wasm32 compiled with panic=abort by default`
or compiling with panic=abort:
 `// compile-flags: -C panic=abort`.

ouz-a commented:
> The tests fail on wasm32-unknown-unknown, because this target defaults to `-C panic=abort`, and the function calls in MIR don't have unwind edges:
> 
> ```diff
> -           (*_5) = f() -> [return: bb2, unwind: bb5]; // scope 0 at $DIR/derefer_inline_test.rs:10:9: 10:12
> +           (*_5) = f() -> bb2;
> ```
> 
> We have been generally either ignoring test on this target: `// ignore-wasm32 compiled with panic=abort by default` or compiling with panic=abort: `// compile-flags: -C panic=abort`.

Thanks, going to add `// compile-flags: -C panic=abort` to the test, I was pretty confused why the test was running fine on my machine and CI but failing on merge...

oli-obk commented:
@bors r=davidtwco

bors commented:
:pushpin: Commit e71913e847666ae9c35417c9590091d72ec6a5e9 has been approved by `davidtwco`

<!-- @bors r=davidtwco e71913e847666ae9c35417c9590091d72ec6a5e9 -->
<!-- homu: {"type":"Approved","sha":"e71913e847666ae9c35417c9590091d72ec6a5e9","approver":"davidtwco"} -->

bors commented:
:hourglass: Testing commit e71913e847666ae9c35417c9590091d72ec6a5e9 with merge c35035cefc709abddabfb28ecc6a326458d46ce2...
<!-- homu: {"type":"BuildStarted","head_sha":"e71913e847666ae9c35417c9590091d72ec6a5e9","merge_sha":"c35035cefc709abddabfb28ecc6a326458d46ce2"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/6661832643?check_suite_focus=true)
Approved by: davidtwco
Pushing c35035cefc709abddabfb28ecc6a326458d46ce2 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"davidtwco","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/6661832643?check_suite_focus=true"},"merge_sha":"c35035cefc709abddabfb28ecc6a326458d46ce2"} -->

rust-highfive commented:
📣 Toolstate changed by rust-lang/rust#97025!

Tested on commit rust-lang/rust@c35035cefc709abddabfb28ecc6a326458d46ce2.
Direct link to PR: <https://github.com/rust-lang/rust/pull/97025>

💔 miri on windows: test-pass → test-fail (cc @oli-obk @eddyb @RalfJung).
💔 miri on linux: test-pass → test-fail (cc @oli-obk @eddyb @RalfJung).

rust-timer commented:
Finished benchmarking commit (c35035cefc709abddabfb28ecc6a326458d46ce2): [comparison url](https://perf.rust-lang.org/compare.html?start=4a8d2e3856c0c75c71998b6c85937203839b946d&end=c35035cefc709abddabfb28ecc6a326458d46ce2&stat=instructions:u).


### [Instruction count](https://perf.rust-lang.org/compare.html?start=4a8d2e3856c0c75c71998b6c85937203839b946d&end=c35035cefc709abddabfb28ecc6a326458d46ce2&stat=instructions:u)

- Primary benchmarks: no relevant changes found
- Secondary benchmarks: 😿 relevant regressions found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | N/A      | N/A | 0         |
| Regressions 😿 <br /> (secondary) | 0.6%     | 0.8% | 3         |
| Improvements 🎉 <br /> (primary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (secondary) | N/A      | N/A | 0         |
| All 😿🎉 (primary) | N/A      | N/A | 0         |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=4a8d2e3856c0c75c71998b6c85937203839b946d&end=c35035cefc709abddabfb28ecc6a326458d46ce2&stat=max-rss)
<details>
<summary>Results</summary>


- Primary benchmarks: 🎉 relevant improvements found
- Secondary benchmarks: 😿 relevant regressions found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | N/A      | N/A | 0         |
| Regressions 😿 <br /> (secondary) | 3.7%     | 5.4% | 4         |
| Improvements 🎉 <br /> (primary) | -2.7%    | -3.2% | 2         |
| Improvements 🎉 <br /> (secondary) | N/A      | N/A | 0         |
| All 😿🎉 (primary) | -2.7%    | -3.2% | 2         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=4a8d2e3856c0c75c71998b6c85937203839b946d&end=c35035cefc709abddabfb28ecc6a326458d46ce2&stat=cycles:u)
<details>
<summary>Results</summary>


- Primary benchmarks: 😿 relevant regression found
- Secondary benchmarks: 🎉 relevant improvement found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | 2.4%     | 2.4% | 1         |
| Regressions 😿 <br /> (secondary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (primary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (secondary) | -2.3%    | -2.3% | 1         |
| All 😿🎉 (primary) | 2.4%     | 2.4% | 1         |
</details>

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression
